### DISCUSSION TITLE
Composition of routers / schema merging and the define-once global variable t

### DISCUSSION THREAD

**Discussion by Jacse** [trpc/trpc#2712] [General]
In GraphQL and express etc. it's possible to compose multiple routers/schemas bottom-up. However with the introduction of the define-once global variable `t` in `v10` this seems impossible to do cleanly without circular dependencies.

Imagine the following scenario. I have services that take care of fetching movies, reviews, and users for IMDB. I want to decouple these, so I've created separate `libs` in my Nx monorepo and I want to combine them all and expose them in my shallow `app` called `api`. With express, I can simply export a router in each `lib` and combine them with namespaces/prefixes in the `api` `app`. How do I achieve something similar with `rRPC`?

The only way I could think of now is to pass the global variable `t` into some sort of constructor in each lib, but that creates a messy circular dependency and makes testing quite hard.

Why do procedures need to be defined on the instantiated `t` variable? If they could be defined without it (like an express router) it would be much easier to create composable/decoupled apps and test them individually.

Maybe I've missed some architectural philosophy or something else, but I'd love to hear your thoughts.

**Comment by KATT** [comment]
I don't see where the circular dependency is created.

The `t` variable can be it's own package in the monorepo that is not importing anything and is used by other "service-packages"

**Reply by Jacse** [reply]
Ah yes of course. I do think it feels a little "off" to create one variable (basically two lines of code) in a lib which must then be  passed to all services in stead of being able to build routers separately, but I understand that's how its done in tRPC. Thanks for the help!

**Reply by mmlngl** [reply]
@Jacse I'm trying to do something similar (also with nx) 

The circular dependency pops-up for me like this:

- `lib/providers/trpc/server.ts` exposes `t` variable
- `lib/contexts/assets` imports `t` variable and defines local assets router
- `lib/contexts/projects` imports `t` variable and defines local projects router

Then, back in `lib/providers/trpc/router.ts`, I try to import the local assets and projects router.
That's when the circular dep appears.

**Reply by Jacse** [reply]
@mmmoli you'll have to have your router be a separate library from your server. In that way they assets and proejcts depend on the server, and the router depends on assets, projects, and server. I my opinion the structure you end up with is pretty unergonomic, but it works.

**Reply by mmlngl** [reply]
@Jacse I get you. Thanks!
